### PR TITLE
cli/sql: report the txn status in prompt

### LINKED ISSUE #10290: cli/sql: feature request: report the txn status in prompt
Currently when a user has a long interactive session it can become confusing/uncertain whether there is a transaction currently ongoing server-side.

The CLI shell could request the current transaction status just before each prompt and use that to adapt the prompt. For example:

```
root@:26257    >SELECT 1;
...
root@:26257    >BEGIN;
BEGIN
root@:26257 TXN> 
root@:26257 TXN> SELECT foo FROM nonexistent;
pq: table 'nonexistent' not found
root@:26257 ERR> 
root@:26257 ERR> SELECT 1;
pq: transaction aborted, ignoring statements
root@:26257 ERR> 
root@:26257 ERR> ROLLBACK;
ROLLBACK
root@:26257    > 
```

To achieve this two steps are needed:
1. a new SHOW variable must probably be introduced to report the current txn status
2. the CLI shell must be extended to issue  SHOW in between prompts, like it does already for SYNTAX.

### PR DESCRIPTION
**Pull request by yznming**
Report the transaction status in prompt.

Fixes #10290. 

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10866)

### CODE REVIEW

**Comment by knz** [comment]
cc @andreimatei what do you think?
I like this approach (a new `PreState` attribute). However there is an alternative: to special case the new `ShowTransactionStatus` statement node in the various `execStmt` methods in the executor. Which one do you think reacts best to txn retries?

**Comment by knz** [comment]
@yznming: please add a test in pkg/cli/interactive_tests to validate that the shell prompt is displayed properly.

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2c_S4ZxeBP-V4VZQV:-KX2c_S597YrCuLwlX1n:b-25bj22) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L365)):*
> ```Go
> 			c.fullPrompt, _ = preparePrompts(c.conn.url, sql.Aborted.String())
> 		} else {
> 			c.fullPrompt, _ = preparePrompts(c.conn.url, sql.NoTxn.String())
> ```

I don't know if that's true that an error causes the transaction status to become "none".
I think the best you can do here is "unknown".

---

*[pkg/cli/sql.go, line 406 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2coBQfIyILf3t9Kj-:-KX2coBR4GveEpkz2tas:b-l1o5k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L406)):*
> ```Go
> 	case sql.Aborted.String():
> 		fullPrompt += " ERR> "
> 	default:
> ```

Why did you omit the case for "transaction in retry state"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KX2d5rGvOyLNTkYO2KJ:b-xkvj3t)*

**Comment by yznming** [comment]
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2c_S4ZxeBP-V4VZQV:-KX3wj8VJULsGFzj5ApY:b-f4122k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L365)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I don't know if that's true that an error causes the transaction status to become "none".
> I think the best you can do here is "unknown".
</details>
if the cli gets into "unknown" state, how to reset the state into "None"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866)*

**Comment by knz** [comment]
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2c_S4ZxeBP-V4VZQV:-KX4coISiEwVK9_aP2OC:b7zibzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L365)):*
<details><summary><i>Previously, yznming (Yan Long) wrote…</i></summary>
> if the cli gets into "unknown" state, how to reset the state into "None"?
</details>
Every time the user presses 'enter', either on an empty prompt or with some non-empty line, the entire state machine will run back to doStartLine and run refreshPrompts() again. I expect there will be a point where the transaction state will become known again. Why don't you try it out? 

(For example, try stopping the cockroach server while the shell is running. You will see that the query to retrieve the txn status will fail. Then start the server again, and press "enter" a few times. I think the txn status will become known again.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866)*

**Comment by yznming** [comment]
done

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cli/sql.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2c_S4ZxeBP-V4VZQV:-KX697vuIKQS1vARhEVl:b-818rbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L365)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Every time the user presses 'enter', either on an empty prompt or with some non-empty line, the entire state machine will run back to doStartLine and run refreshPrompts() again. I expect there will be a point where the transaction state will become known again. Why don't you try it out? 
> 
> (For example, try stopping the cockroach server while the shell is running. You will see that the query to retrieve the txn status will fail. Then start the server again, and press "enter" a few times. I think the txn status will become known again.)
</details>
done. I think the prompts should be refreshed in `doProcessingFirstLine`.  bacause an empty prompt would not get into `doStartLine`.

---

*[pkg/cli/sql.go, line 406 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2coBQfIyILf3t9Kj-:-KX6A3yMjAH-8B83-vGf:b-hcz9xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L406)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Why did you omit the case for "transaction in retry state"?
</details>
the `RestartWait` state should be shown as `TXN>`, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KX6Asq1R79kmpW4f41O:bb7f2mf)*

**Comment by andreimatei** [comment]
I wouldn't put the `PreState` on the `txnState` - seems like a random field that's not actually "state", it's something we re-initialize for each statement. How about we put it in the more transient `planner.evalCtx`?
Better yet, how about we replace it with a `evalCtx.implicitTxn`? There's already an `implicitTxn` being computed and passed around, so maybe we can move that to the `evalCtx`. Then, the state returned by `SHOW STATE` would be `txnState.State`, unless `evalCtx.implicitTxn` is set, in which case we'd return `NoTxn`.

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/session.go, line 231 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX7-Wge0ieOfOaC3c-r:-KX7-Wgfa0Ge71stT6AY:b-yrhb55) ([raw file](https://github.com/cockroachdb/cockroach/blob/698ea705cd2090d3de9d48ef62a3beeabdbb0ef9/pkg/sql/session.go#L231)):*
> ```Go
> type txnState struct {
> 	txn      *client.Txn
> 	PreState TxnStateEnum
> ```

if we keep this member, please add a comment describing it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KX70NihtrPEmmutvlYl:bq8334g)*

**Comment by knz** [comment]
@yznming: see my comments and Andrei's. 

Again thank you for working on this. It is a bit more complicated than I initially envisioned but you are on a good path.

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 365 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2c_S4ZxeBP-V4VZQV:-KX8O8EDkujRJNUT1qYT:b-qejzwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L365)):*
<details><summary><i>Previously, yznming (Yan Long) wrote…</i></summary>
> done. I think the prompts should be refreshed in `doProcessingFirstLine`.  bacause an empty prompt would not get into `doStartLine`.
</details>
That's ok, you can update the state machine accordingly.

---

*[pkg/cli/sql.go, line 406 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2coBQfIyILf3t9Kj-:-KX8OEF4pXUBFvSV6xTp:b-279hql) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L406)):*
<details><summary><i>Previously, yznming (Yan Long) wrote…</i></summary>
> the `RestartWait` state should be shown as `TXN>`, right?
</details>
I think it is useful to separate the states visually. 
In the RestartWait state the user cannot send the same statements, and that is important to communicate.

Also I am fond of same-width identifiers for all the states, so that the prompt `>` is always on the same column on the screen.

Can you try to find names and alignments that work well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KX8Obju56ox9_BOHJHN:bgsoom4)*

**Comment by yznming** [comment]
@knz what do you think about using transaction error to detect `restartwait` and `commitwait` states?

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXB6cVPveLqn5lQXTMJ:b-rf6445)*

**Comment by knz** [comment]
The problem is that if you issue any statement other than ROLLBACK TO SAVEPOINT or SAVEPOINT in the retrywait state, then the retrywait state is lost entirely (goes to the aborted state).

It is necessary I think to add a special case in the method `execStmtInAbortedTxn` in executor.go.

**Comment by yznming** [comment]
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 406 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KX2coBQfIyILf3t9Kj-:-KXE--wzVUpHxTCUIYvk:b-fycr7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c49b6f981fff8cc20e9cd9f78a1aa8f5eccfaed/pkg/cli/sql.go#L406)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think it is useful to separate the states visually. 
> In the RestartWait state the user cannot send the same statements, and that is important to communicate.
> 
> Also I am fond of same-width identifiers for all the states, so that the prompt `>` is always on the same column on the screen.
> 
> Can you try to find names and alignments that work well?
</details>
@knz how about `INTXN`, `ERROR`, `CWAIT`, `RWAIT`, `UNKWN`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866)*

**Comment by knz** [comment]
I find these names too complicated to read. Also perhaps I was wrong and different widths are ok, as long as there are enough spaces on the left to align them on the screen. For example:

```
     > (all spaces: NoTxn)
 OPEN> (Open)
ERROR> (Aborted + Error)
RETRY> (Aborted + RetryWait)
 DONE> (CommitWait)
?????> (unknown)
````

**Comment by yznming** [comment]
@knz got it. by the way, do you have any idea how to make a retriable scenario?

**Comment by knz** [comment]
I'm going to investigate how to trigger a retry error for you.

**Comment by knz** [comment]
@yznming please use the patch in #11193 to test your contribution (you will need to pull from my branch manually until that PR gets merged).

**Comment by yznming** [comment]
@knz done

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXE0IopHxA9munwUAVQ:b-bxjwka)*

**Comment by knz** [comment]
I am impressed, this code is very good and it works well. 
Just a minor suggested improvement.

Also there is one issue remaining with the server reporting a warning in the Aborted state, and I think the logic to refresh the prompts can be improved. For this I make a couple suggestions here: https://github.com/yznming/cockroach/pull/1
If you are OK with this you can accept this PR, then the changes will show up automatically here.
Feel free to squash my commits into yours, so that the git history stays simple.

---

Reviewed 1 of 5 files at r2, 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1087 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KXKQ14MgCv0A8EamN9W:-KXKQ14MgCv0A8EamN9X:bg9b1ms) ([raw file](https://github.com/cockroachdb/cockroach/blob/60253f6389a3d18a4f0ac9cf98d944c9903d0846/pkg/sql/executor.go#L1087)):*
> ```Go
> 
> const (
> 	stmtTxnStatus = `TRANSACTION STATUS`
> ```

Remove this (see my other comment)

---

*[pkg/sql/parser/sql.y, line 1515 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-KXKQ4qI8YVtpGOVKk40:-KXKQ4qI8YVtpGOVKk41:bnom900) ([raw file](https://github.com/cockroachdb/cockroach/blob/60253f6389a3d18a4f0ac9cf98d944c9903d0846/pkg/sql/parser/sql.y#L1515)):*
> ```Yacc
> | SHOW TRANSACTION STATUS
>   {
>     $$.val = &Show{Name: "TRANSACTION STATUS"}
> ```

Please create a new struct `ShowTxnStatus` without attributes (like `ShowUsers`) and use it as special case in `executor.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXKPqNHXsamRGzTVlGb:b-h7rlm3)*

**Comment by knz** [comment]
Don't forget to squash the commits into one; also my proposed fix assumes there is a new `ShowTxnStatus` statement but you haven't implemented that yet (as far as I can see).

---

Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXKml9TSXZ0nn8ZQECj:bps7b2y)*

**Comment by knz** [comment]
To squash: do `git rebase -i master` then select the option "squash" on all the commits after the first one.

**Comment by yznming** [comment]
go it. thank you for reminding me.

**Comment by yznming** [comment]
@knz done

File: pkg/sql/executor.go [hunk 1]
```diff
@@ -1078,6 +1091,25 @@ func (e *Executor) noTransactionHelper(txnState *txnState) (Result, error) {
 	return Result{Err: errNoTransactionInProgress}, errNoTransactionInProgress
 }
 
+// showTransactionState returns the state of current transaction.
+func showTransactionState(txnState *txnState, implicitTxn bool, planMaker *planner) (Result, error) {
```
Comment (knz) [inline pkg/sql/executor.go]: 1) rename to `runShowTransactionState`
2) make this function a method of planner: `func (p *planner) runShowTransactionState(...)`

File: pkg/sql/executor.go [hunk 2]
```diff
@@ -783,25 +791,30 @@ func (e *Executor) execStmtsInCurrentTxn(
...
-		default:
-			panic(fmt.Sprintf("unexpected txn state: %s", txnState.State))
-		}
-		if (e.cfg.TestingKnobs.CheckStmtStringChange && false) ||
-			(e.cfg.TestingKnobs.StatementFilter != nil) {
-			if after := stmt.String(); after != stmtStrBefore {
-				panic(fmt.Sprintf("statement changed after exec; before:\n    %s\nafter:\n    %s",
-					stmtStrBefore, after))
+		if _, ok := stmt.(*parser.ShowTransactionStatus); ok {
+			res, err = showTransactionState(txnState, implicitTxn, planMaker)
```
Comment (knz) [inline pkg/sql/executor.go]: Add a comment here:
```
// Run SHOW TRANSACTION STATUS in a separate code path so it is
// always guaranteed to execute regardless of the current transaction state.
```

**Review by knz — COMMENTED** [review]
Thanks, just a minor change left below.

**Comment by yznming** [comment]
@knz done

**Comment by knz** [comment]
This looks great.
:lgtm:

---

Reviewed 5 of 6 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXKmukuoi189BfuRKmt:b2jzffw)*

**Comment by knz** [comment]
@bdarnell can you have a cursory look and see if the idea works for you?

**Comment by bdarnell** [comment]
LGTM

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10866#-:-KXQjDn6GrTAlJxd2Rg1:bb6ywmx)*

**Comment by knz** [comment]
@yznming thank you very much for your help!

**Pull request merged**